modules/calc/src/main/java/com/opengamma/strata/calc/Results.java

```diff
@@ -207,17 +219,31 @@ private String invalidColumnNameMessage(ColumnName columnName) {
    * <p>
    * The result may be a single value or a multi-scenario value.
    * A multi-scenario value will implement {@link ScenarioArray} unless it has been aggregated.
+   * <p>
+   * If the calculation did not complete successfully, a failure result will be returned
+   * explaining the problem. Callers must check whether the result is a success or failure
+   * before examining the result value.
    *
    * @param <T>  the result type
    * @param rowIndex   the index of the row containing the results for a target
    * @param columnName  the name of the column
    * @param type  the result type
-   * @return the results for the specified row and column for a set of scenarios
+   * @return the result for the specified row and column for a set of scenarios, cast to the specified type
    * @throws IllegalArgumentException if the row index or column name is invalid
    * @throws ClassCastException if the result is not of the specified type
    */
   public <T> Result<T> get(int rowIndex, ColumnName columnName, Class<T> type) {
-    return get(rowIndex, columnName).map(v -> type.cast(v));
+    return cast(get(rowIndex, columnName), type);
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> Result<T> cast(Result<?> result, Class<T> type) {
+    // cannot use result.map() as we want the exception to be thrown
+    if (result.isFailure() || type.isInstance(result.getValue())) {
```

Comment:
`Result.getValue` returns a `T` so a cast method isn't needed. Just cast the value:

    Result<?> result = ...;
    String str = (String) result.getValue();

is just as good as

    Result<?> result = ...;
    String str = result.getValue(String.class);

and doesn't require an extra method.

It's valuable here because `CalculationResult` returns `Result<T>` so there's no way to get a `Result<Foo>` from a `CalculationResult<?>` without an unchecked cast.